MHub / GC - Add the GC Tiger challenge winner LB2 model

silvandeleemput opened the pull request:
It took a bit longer than expected, but this PR adds the GC Tiger challenge winner LB2 model to MHub. 
* This implements a Dockerfile based on the new base image
* I added two pipeline files:
    * `default.yml/run.py` (WSI-DICOM in JSON out) <--- This one uses the PanImgConverter
    * `tiff_pipeline.yml/run_tiff_pipeline.py` (TIFF in JSON out) <-- This one was added for convenience/comparison/testing
* The PR should probably be retargeted to an `m-gc-tiger-lb2` branch, subsequently, the MHub model repo should be properly added to the Dockerfile

This is also an interesting model to have a look at because of the following issues/characteristics:
  * Uses WSI as an input and uses the PanImgConverter
  * Has a JSON output with a single integer value, which is the TIL score. 
  * The TigerLB2Runner requires a GPU to be run (enforced through an assert, otherwise it would throw an error). It will require elaborate hacking to get working on a CPU (I have been able to do it for a CPU-only version), but the runtime is over one hour for the CPU version versus 15-20 mins on the GPU so I think not worth supporting it. Let me know otherwise.
  * ~~This model suffers from the model import issue, (check the Dockerfile), which required me to perform some ugly hacks that even prevent me from running the pipeline directly from the configuration files.~~

silvandeleemput committed:
add initial nocuda Dockerfile and working TiffImporter and run.py script

silvandeleemput committed:
add cuda 11.4/12.0 Dockerfiles and replaced print statement in runner

silvandeleemput committed:
cleanup ASAP install dockerfile line and add matching torchvision versions to torch install

silvandeleemput committed:
Merge branch 'MHubAI:main' into m-tiger-lb2

silvandeleemput committed:
moved tiger_lb2 -> gc_tiger_lb2, changed Dockerfile to new base img, fixed model weights install, updated pipelines

silvandeleemput committed:
add DicomImporter import to run.py

silvandeleemput committed:
add torch cuda assert to tiger_lb2 runner

silvandeleemput committed:
Added a fix for the import issue by introducing a CLI and a subprocess call from the pipeline

* Fix for the import issue:
  * Implemented a CLI for the Tiger lb2 runner that takes an input TIFF path and output JSON file path
  * Integrated the CLI inside the TigerLB2Runner as a subprocess call to the CLI through a fresh Python interpreter
  * Cleaned the Dockerfile some more

* Caveats:
  * Because of the fix the extra run script for the tiff_pipeline became redundant, can be called directly from the tiff_pipeline.yml through a call of mhubio.run
  * For the main pipeline the run.py is still required since the TiffPanImgConverter is not automatically picked up

* Updated the PanImgConverters, since the returned UnconsumedFilesException e.errors were renamed to e.file_errors

silvandeleemput commented:
Pushed a fix for the model import issue as suggested by @LennyN95 to this PR. 
To circumvent the conflicting imports issue I added a CLI script for running the model algorithm that is now called as a subprocess within the pipeline. I consider the issue solved for this particular model.

LennyN95 commented:
Can we use ValueOutput+ReportExporter for this algorithm? 
If the JSON generation is characteristic to the algorithm, I suggest keeping the JSON file as a data output and adding a ValueOutput with the score for greater flexibility and compatibility. This way, we allow users to either use the original file and simultaneously offer full flexibility via our ReportExporter mechanisms.

silvandeleemput committed:
Merge branch 'MHubAI:main' into m-gc-tiger-lb2

silvandeleemput committed:
Update and cleanup the tiger lb2 algorithm model code

Dockerfile
  - fixed to commit for tiger_challenge repo
  - updated MHubAI/models clone stub
  - updated entrypoint

pipeline default.yml
  - updated the outputs
  - use the TiffConverter
  - Added reportexporter template

runner:
  - Added TIL Score as ValueOutput
  - Confirmed data output
  - Added/updated comments

removed:
  - custom panimg converter classes
  - run.py script to run the pipeline

silvandeleemput committed:
Merge branch 'm-gc-tiger-lb2' of github.com:DIAGNijmegen/MHubAI_models into m-gc-tiger-lb2

silvandeleemput commented:
This PR has been cleaned and updated to meet the latest requirements:
- Dockerfile
  - fixed to commit for tiger_challenge repo
  - updated MHubAI/models clone stub
  - updated entrypoint
- pipeline default.yml
  - updated the outputs
  - use the TiffConverter
  - added reportexporter settings
- runner:
  - Added TIL Score as ValueOutput
  - Confirmed data output
  - Added/updated comments
- removed:
  - custom panimg converter classes
  - run.py script to run the pipeline

> Can we use ValueOutput+ReportExporter for this algorithm? If the JSON generation is characteristic to the algorithm, I suggest keeping the JSON file as a data output and adding a ValueOutput with the score for greater flexibility and compatibility. This way, we allow users to either use the original file and simultaneously offer full flexibility via our ReportExporter mechanisms.

@LennyN95 I added a ValueOutput for the TIL score for this algorithm, but when I try to export that value with the ReportExporter it outputs an empty report with: `{}` I get the following error using the `--print` option: 

```
Start ReportExporter
Error while generating report for  {'data': 'til_score', 'label': 'TIL score', 'value': 'value'}
```

After doing some in-situ debugging, I found that the following line fails:

`mhubio/modules/exporter/ReportExporter.py::153:  assert len(data) == 1, "data name not unique."` 

Then I found out that `data = []` is an empty list after printing the data, which was populated using the following code:
```
# fetch the data
data = [d for d in instance.outputData if d.name == include['data']]
```

It appears that `instance.outputData = []` is also empty, so it seems that the outputData is somehow not properly picked up when using this decorator in the TigerLB2Runner class:

`@IO.OutputData('til_score', TilScoreOutput, data='in_data', the='TIGER LB2 TIL score - percentage of stromal area covered by tumour infiltrating lymphocytes. Values between 0-100 (percent).')`


@LennyN95 Do you know what is going on here? The latest code update in this PR is the code that I used. I only disabled the ReportExporter in the default.yml for now to make this fully functional, but during testing this was enabled. 

I got it working with similar code before for other model PRs. It doesn't however seem to be caused by codebase changes, because I tried to revert it to older versions of mhubio up until August 3th, but it still had the same output and error. So I guess it is something else.

silvandeleemput committed:
fix issue with last assert runner, re-enable ReportExporter

silvandeleemput commented:
Alright, I got it working, this PR can now use the ReportExporter as requested. 

Apparently it was a mistake in the last `assert` line of the runner (called the wrong method). The error message was reported in the output using `--print`, but because the code doesn't stop at the first error, the rest of the output obfuscated this error message. I think this further emphasizes the need for a `--stop-on-error` flag to avoid unnecessary debugging.

silvandeleemput committed:
replace subprocess call with self.subprocess, replace asserts with explicit errors

silvandeleemput committed:
add basic meta.json

silvandeleemput committed:
add minimum panimg pip install fix for wsi dicom conversion

silvandeleemput committed:
changed algorithm details type

LennyN95 commented on models/gc_tiger_lb2/utils/TigerLB2Runner.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@ValueOutput.Name('til_score')
+@ValueOutput.Meta(Meta(key="value"))
```
This is a placeholder.
Consider adding valuable meta-data such as the value format (e.g., scale=percent, min=0, max=100)

silvandeleemput replied:
Ok, modified.

LennyN95 commented on models/gc_tiger_lb2/utils/TigerLB2Runner.py:
```diff
@@ -0,0 +1,59 @@
+            ]
+        )
+
+        if not Path(out_data.abspath).is_file():
```
This is not required; our `IO.Output` decorator performs a file-existence check and confirms only existing files using the official data API (`data.confirm()`).

LennyN95 replied:
If required for a subsequent step, pls invert the expression instead of throwing exceptions/errors.

silvandeleemput replied:
I think doing it this way is preferable since the raised error clearly communicates what the issue is within the module. If the expression is inverted it would become a silent failure, which is much harder to debug.

LennyN95 replied:
`Something went wrong` does not at all `clearly communicate what the issue` is.  
Either improve or update as suggested.

silvandeleemput replied:
I'll change the error message to a better one.

LennyN95 commented on models/gc_tiger_lb2/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+RUN pip3 install --no-cache-dir --force-reinstall numpy==1.22
+
+# Enforce minimum version of panimg (with WSI fix)
+RUN pip3 install --no-cache-dir panimg>=0.13.2
```
Did you encounter any issues without specifying this line? 
The panimg from our base image should already satisfy the requirement.

silvandeleemput replied:
Yes, there was a fix in that specific version, and wasn't part of the base image at the time, it is redundant now and I removed it.

LennyN95 commented on models/gc_tiger_lb2/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+RUN pip3 install --no-cache-dir -r /vuno/requirements.txt
+
+# Reinstall correct version of Numpy to function with ASAP 2.1
+RUN pip3 install --no-cache-dir --force-reinstall numpy==1.22
```
This might clash with future mhubio updates. One option as once discussed could be wrapping up all algorithm dependencies in a pip env and then calling the cli script in that context.

silvandeleemput replied:
Could be, but the other modules besides the runner used for this model don't use Numpy as far as I can tell correct? I rather not touch this, since this will be very hard to configure correctly in combination with ASAP.

LennyN95 replied:
Will you go back and fix it the moment we use Numpy in e.g. calculate general run-time statistics :)?
I already made a suggestion, you can install ASAP and the fixed Numpy version in a pip-env.

Please update before we can proceed.

silvandeleemput replied:
I'll try to install them in a pip-env, it'll give some overhead, but less maintainability issues.

silvandeleemput replied:
It has been installed inside a pipenv.

LennyN95 commented on models/gc_tiger_lb2/config/default.yml:
```diff
@@ -0,0 +1,31 @@
+    import_dir: sorted_data
+    sort_data: True
+    meta: 
+      mod: sm
```
`mod: "%Modality"` can be used to automatically parse the dicom Modality header.

silvandeleemput replied:
I rather keep it at `sm`, since that should be the correct DICOM modality anyway for this pipeline to run as all inputs should be slide microscopy images. Is there any benefit to changing it here, except that nothing runs if it does not match the modality?

LennyN95 replied:
Yes, e.g. Patient folder contains two dicom series, one SM and the other CT. It makes no sense to import both as SM. 

Please change as suggested.

silvandeleemput replied:
Ok, I'll update it as suggested.

LennyN95 commented on models/gc_tiger_lb2/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+    && apt-get -f install --fix-missing --fix-broken --assume-yes \
+    && ldconfig -v \
+    && apt-get clean \
+    && echo "/opt/ASAP/bin" > /usr/local/lib/python3.8/dist-packages/asap.pth \
```
Looks a bit hacky. Is this the official / stable install?

silvandeleemput replied:
Yes, this code came with the Dockerfile used for the Tiger challenge.

LennyN95 replied:
Note, that if the Ubuntu or the Python version changes, this model will become deprecated or require a revision.

Do you know if ASAP can run on Ubuntu 22.04 or with any other Python version?
Did you think about a maintenance strategy?

silvandeleemput replied:
Hi, yes as discussed: There is a newer version of ASAP for Ubuntu 22.04, so once the base image moves, we can move along. The maintenance strategy is to support these Ubuntu and Python versions (potentially in a pyenv), once 22.04 is rolled out we can update accordingly. For steps beyond that we have to discuss...

silvandeleemput replied:
Ok, it is now part of a pipenv.

LennyN95 commented on models/gc_tiger_lb2/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,54 @@
+    unzip /vuno/pretrained_weights.zip -d /vuno && \
+    rm /vuno/pretrained_weights.zip
+
+# Clone the main branch of MHubAI/models TODO
```
To avoid disabling the model import during development, please use the recently introduced model import build utility instead:

```
# Import the MHub model definiton
ARG MHUB_MODELS_REPO
RUN buildutils/import_mhub_model.sh gc_tiger_lb2 ${MHUB_MODELS_REPO}
```

silvandeleemput replied:
Done!

silvandeleemput committed:
Merge branch 'MHubAI:main' into m-gc-tiger-lb2

silvandeleemput committed:
add import mhub model definition lines, remove panimg pip install, remove first line Dockerfile

silvandeleemput committed:
modified meta line

LennyN95 commented on models/gc_tiger_lb2/meta.json:
```diff
@@ -0,0 +1,113 @@
+    },
+    "evaluation": {
+      "title": "Evaluation data",
+      "text": "The test set consists of a separate dataset of n=707 H&E-stained whole-slide breast cancer histopathology images",
```
Is the mentioned dataset closed-source? Please specify or provide a public reference.#
Why is evaluation `summary.data.evaluation.vol_samples = 58 << n = 707`?

silvandeleemput replied:
We will update this.

silvandeleemput replied:
This was changed to the correct value.

LennyN95 commented on models/gc_tiger_lb2/meta.json:
```diff
@@ -0,0 +1,113 @@
+  "name": "tiger_lb2",
+  "title": "TIGER challenge winner: Team VUNO",
+  "summary": {
+    "description": "Participants in the TIGER challenge will have to develop computer algorithms to analyze H&E-stained whole-slide images of breast cancer histopathology, to perform three tasks:  detection of lymphocytes and plasma cells, which are the main types of cells considered as tumor-infiltrating lymphocytes; segmentation of invasive tumor and tumor-associated stroma, which are the main  ...
```
This is misleading and reads like this is an ongoing challenge. It is not clear that this is about a particular (winning) model from a past challenge.

silvandeleemput replied:
We will update this.

silvandeleemput replied:
We have rewritten it to be more like a description of the algorithm.

silvandeleemput committed:
meta.json - fix typo and match model name

silvandeleemput committed:
fix dynamic modality in default.yml, improve error message in runner #38

silvandeleemput committed:
meta.json - change version to 0.1.0

silvandeleemput committed:
Merge branch 'MHubAI:main' into m-gc-tiger-lb2

silvandeleemput committed:
meta.json - modified, updated, and extended

silvandeleemput committed:
Dockerfile - add reproducibility fix

LennyN95 commented on models/gc_tiger_lb2/dockerfiles/Dockerfile:
```diff
@@ -20,9 +20,11 @@ RUN apt-get update \
     cd /vuno && git reset --hard 720f8dfca4624792c8e57915c4222efec5a0c2d4 && \
-    rm -rf /vuno/.git
+    rm -rf /vuno/.git && \
+    sed -i "238i \        torch.manual_seed(0)" /vuno/segmentation/inference.py
```
This is great but should be implemented in the source repository.

silvandeleemput replied:
Yes, ideally we would be able to make the necessary changes in the original repository. However, since it is unlikely that we have that option, we need to discuss the best approach to take. One possible solution is forking the repository and implementing the changes in the forked version. Let's discuss this in more detail during our next meeting. For now, let's focus on the remaining changes and come up with a solid plan to address this issue as soon as possible.

silvandeleemput replied:
I have forked the repository and added the change there. The manual correction is no longer in the Dockerfile.

silvandeleemput committed:
Dockerfile, runner, cli - added pipenv for model algorithm #38

silvandeleemput commented:
[output.zip](https://github.com/MHubAI/models/files/14526578/output.zip)

silvandeleemput commented:
/test

```yaml
sample:
  idc_version: 17.0
  data:
    - SeriesInstanceUID: 1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0
      aws_url: s3://idc-open-data/20ab74ef-080b-4975-adad-7507b7b04cb8/*
      path: dicom

reference:
  url: https://github.com/MHubAI/models/files/14526578/output.zip
```

---

<details><summary>Test Results (24.03.13_17.52.53_6VEUMKpYVb)</summary>

```yaml
id: 7654bec5-0c56-43b5-82db-cee986e4d663
date: '2024-03-13 18:08:50'
checked_files:
- file: gc_tiger_lb2_til_score.json
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_til_score.json
  checks:
  - checker: DataFileCheck
    findings:
    - label: Value similar
      description: Value of key 'TIL score' is similar in source file
      subpath: TIL score
      info:
        scale: 1
        precision: -1
        src: 4
        ref: 5
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DataFileCheck:
      files: 1
      findings:
        Value similar: 1
    SizeCheck:
      files: 1
conclusion: false
```

</details>

silvandeleemput commented:
/review

We have implemented the requested changes and updated the meta.json. 
Except for the reproducibility fix it should be as good as ready.

silvandeleemput commented:
@LennyN95 thanks for checking. Small differences in outcome are to be expected when run on different GPUs due to accumulated small numerical differences. If the results only differ by a few points (something in the range of 1-2 points) this is deemed acceptable for this model as discussed with the challenge organizers.

silvandeleemput committed:
pull source from forked repo, added segmentation output as mha

silvandeleemput committed:
meta.json - add tumor and stroma segmentation mask description to outputs

silvandeleemput commented:
[gc_tiger_lb2_output.zip](https://github.com/MHubAI/models/files/15091074/gc_tiger_lb2_output.zip)

silvandeleemput commented:
@LennyN95 I have created a fork of the source model/algorithm repository here: https://github.com/DIAGNijmegen/tiger_vuno_algorithm. 

In this fork, I have added the reproducibility fix and added the tumor and stroma segmentation mask to the output.
The PR has been updated to build from the fork and output the segmentation.

I have uploaded a new output.zip file (which includes the segmentation) and updated the URL in the test above.

Please check if everything is in order now. I suspect that the segmentation will be different due to GPU hardware/driver differences.

LennyN95 commented:
Thank you @silvandeleemput. I create this as a new test and reverted the modification to keep the test history clean. Test requests should be edited to correct syntax errors only but not touched once one or more tests were performed (and reported in the test request).

LennyN95 commented:
/test

```yaml
sample:
  idc_version: 17.0
  data:
    - SeriesInstanceUID: 1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0
      aws_url: s3://idc-open-data/20ab74ef-080b-4975-adad-7507b7b04cb8/*
      path: dicom

reference:
  url: https://github.com/MHubAI/models/files/15091074/gc_tiger_lb2_output.zip
```

---

<details><summary>Test Results (24.04.25_08.24.29_0ZZ5RYzDlk)</summary>

```yaml
id: fa0fa712-4d2b-4f13-adae-56cc312c4209
date: '2024-04-25 08:42:26'
missing_files:
- 1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_segmentation.mha
checked_files:
- file: gc_tiger_lb2_til_score.json
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_til_score.json
  checks:
  - checker: DataFileCheck
    findings:
    - label: Value similar
      description: Value of key 'TIL score' is similar in source file
      subpath: TIL score
      info:
        scale: 1
        precision: -1
        src: 4
        ref: 5
  - checker: SizeCheck
summary:
  files_missing: 1
  files_extra: 0
  checks:
    DataFileCheck:
      files: 1
      findings:
        Value similar: 1
    SizeCheck:
      files: 1
conclusion: false
```

</details>

---

<details><summary>Test Results (24.04.25_14.08.51_xHF1bsAg8N)</summary>

```yaml
id: 1c55ef8c-8325-4838-898e-63a7ef2b04fe
date: '2024-04-25 14:32:07'
checked_files:
- file: gc_tiger_lb2_segmentation.mha
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_segmentation.mha
  checks:
  - checker: ImageFileCheck
    notes:
    - label: Data Type
      description: Data type of the reference image
      info: uint8
    findings:
    - label: Exception
      description: An exception occurred during check
      info: 'could not convert string to float: '''''
- file: gc_tiger_lb2_til_score.json
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_til_score.json
  checks:
  - checker: DataFileCheck
    findings:
    - label: Missing key
      description: Key 'TIL score' is missing in source file
      subpath: TIL score
  - checker: SizeCheck
    findings:
    - label: Size Difference
      description: file size is smaller than reference
      info:
        src_size: 2
        ref_size: 22
        diff_size: -20
summary:
  files_missing: 0
  files_extra: 0
  checks:
    ImageFileCheck:
      files: 1
      findings:
        Exception: 1
    DataFileCheck:
      files: 1
      findings:
        Missing key: 1
    SizeCheck:
      files: 1
      findings:
        Size Difference: 1
conclusion: false
```

</details>

---

<details><summary>Test Results (24.06.03_10.26.45_9t3n5BDtdf)</summary>

```yaml
id: bc76f65a-b934-4712-a8cf-2c5263f357ae
date: '2024-06-03 10:54:38'
checked_files:
- file: gc_tiger_lb2_segmentation.mha
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_segmentation.mha
  checks:
  - checker: ImageFileCheck
    notes:
    - label: Data Type
      description: Data type of the reference image
      info: uint8
    findings:
    - label: Exception
      description: An exception occurred during check
      info: 'could not convert string to float: '''''
- file: gc_tiger_lb2_til_score.json
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_til_score.json
  checks:
  - checker: DataFileCheck
    findings:
    - label: Value similar
      description: Value of key 'TIL score' is similar in source file
      subpath: TIL score
      info:
        scale: 1
        precision: -1
        src: 4
        ref: 5
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    ImageFileCheck:
      files: 1
      findings:
        Exception: 1
    DataFileCheck:
      files: 1
      findings:
        Value similar: 1
    SizeCheck:
      files: 1
conclusion: false
```

</details>

---

<details><summary>Test Results (24.06.07_12.29.30_uTzBQF9Njk)</summary>

```yaml
id: 3302c113-2efa-45e3-97db-7e651d4f835f
date: '2024-06-07 12:46:36'
checked_files:
- file: gc_tiger_lb2_segmentation.mha
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_segmentation.mha
  checks:
  - checker: ImageFileCheck
    notes:
    - label: Data Type
      description: Data type of the reference image
      info: uint8
    - label: Dice Score
      description: Dice score between reference and test image
      info: 1.0
- file: gc_tiger_lb2_til_score.json
  path: /app/test/src/1.3.6.1.4.1.5962.99.1.1334438926.1589741711.1637717011470.2.0/gc_tiger_lb2_til_score.json
  checks:
  - checker: DataFileCheck
    findings:
    - label: Value similar
      description: Value of key 'TIL score' is similar in source file
      subpath: TIL score
      info:
        scale: 1
        precision: -1
        src: 4
        ref: 5
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    ImageFileCheck:
      files: 1
    DataFileCheck:
      files: 1
      findings:
        Value similar: 1
    SizeCheck:
      files: 1
conclusion: false
```

</details>

silvandeleemput committed:
runner.py - fix output type filename

silvandeleemput commented:
@LennyN95 Regarding the results, the exact same exception as last time shows:
```    notes:
    - label: Data Type
      description: Data type of the reference image
      info: uint8
    findings:
    - label: Exception
      description: An exception occurred during check
      info: 'could not convert string to float: '''''
```

I don't understand the origin of this error yet and it wasn't present two base images ago. I will investigate this further.

silvandeleemput commented:
@LennyN95 The results look fine now.
